design-pattern.md

```diff
@@ -429,3 +429,123 @@ public class Main {
+
+Si noti che `Strategy` e `ConcreteStrategy` sono classi che hanno un solo metodo, sono quindi **interfacce funzionali**. Possiamo quindi usare le **espressioni lambda** per definire le `ConcreteStrategy` senza doverle creare esplicitamente.
+
+#### Esempio (senza interfacce funzionali)
+
+Supponiamo di voler gestire un sistema di apertura di casse in un gioco. le casse possono dare, una volta aperte, degli effetti al giocatore che le apre; ad esempio, una cassa può curare il giocatore, un'altra può avvelenarlo, ecc. 
+Usiamo uno **Strategy**.
+
+Per prima cosa, definiamo l'interfaccia `Strategy`, in questo esempio la chiamiamo `EffettoCassa`, che definisce il metodo `apri()`:
+
+```java
+public interface EffettoCassa {
+  public void applicaEffetto();
+}
+```
+
+Definiamo le varie classi `ConcreteStrategy`, in questo esempio le chiamiamo `EffettoCassaCurativa`, `EffettoCassaAvvelenata`, ecc., che implementano `EffettoCassa`:
+
+```java
+public class EffettoCassaCurativa implements EffettoCassa {
+  @Override
+  public void applicaEffetto() {
+    System.out.println("Sei stato curato!");
+  }
+}
+```
+
+```java
+public class EffettoCassaAvvelenata implements EffettoCassa {
+  @Override
+  public void applicaEffetto() {
+    System.out.println("Sei avvelenato!");
+  }
+}
+```
+
+Definiamo la classe `Context`, in questo esempio la chiamiamo `ApriCassa`, che usa un `EffettoCassa` _generico_ e chiama il metodo `apri()`:
+
+```java
+public class ApriCassa {
+  private EffettoCassa effettoCassa;
+
+  public ApriCassa(EffettoCassa effettoCassa) {
+    this.effettoCassa = effettoCassa;
+  }
+
+  public void apri() {
+    effettoCassa.applicaEffetto();
+  }
+}
+```
+
+Il Main poi assegnerà una `ConcreteStrategy` (`EffettoCassaCurativa`, `EffettoCassaAvvelenata`, ecc.) a quella _generica_ (`EffettoCassa`) nel costruttore di `ApriCassa`.
+
+#### Esempio (con interfacce funzionali)
+
+Usando le interfacce funzionali, possiamo usare una sola classe `Cassa`:
+
+```java
+public class Cassa {
+  Runnable cassa;
+
+  public Cassa(Runnable cassa) {
+    this.cassa = cassa;
+  }
+
+  public void apri() {
+    cassa.run();
+  }
+}
+```
+
+Per il nostro uso, definiamo una `Runnable`, che è un'interfaccia funzionale che ha un solo metodo: `run()` per eseguire il codice. A seconda dei tipi dei parametri e del tipo del valore restituito, possiamo usare anche le altre interfacce funzionali spiegate nel capitolo "Programmazione funzionale".
+
+Il `Main` può quindi definire le `Runnable` _concrete_ tramite espressioni lambda e, richiamando il costruttore di `Cassa`, assegnarle a quella _generica_ per eseguire il giusto comportamento:
+
+```java
+public class Main {
+  public static void main(String[] args) {
+    // Crea una cassa curativa
+    Cassa cassaCurativa = 
+    new Cassa(() -> {
+      System.out.println("Sei stato curato!");
+    });
+    /* equivalente a:
+    Runnable cassaCurativa = () -> {
```

Comment:
Cambierei `Runnable cassaCurativa` in `Runnable effettoCassaCurativa`.
Anche perché, dato che sotto `cassaCurativa` viene ridefinita (ed in quel caso il nome è appropriato essendo un oggetto che rappresenta una cassa), non credo ti faccia neanche compilare